components/brave_sync/brave_profile_sync_service_impl.cc

```diff
@@ -305,6 +297,20 @@ void BraveProfileSyncServiceImpl::OnDeleteDevice(const std::string& device_id) {
     const std::string object_id = device->object_id_;
     SendDeviceSyncRecord(SyncRecord::Action::A_DELETE, device_name, device_id,
                          object_id);
+    if (device_id == brave_sync_prefs_->GetThisDeviceId()) {
+      // Give a chance to send the record, because setting sync enabled to
+      // false unloads the extension. So after 1 sec emualate
+      // OnResolvedPreferences response to shutdown even if we are offline.
+      // In worse case we will just have double call of ResetSyncInternal.
+      RecordsListPtr records = CreateDeviceOperationRecord(
+          device_name, object_id, SyncRecord::Action::A_DELETE, device_id);
+      base::PostDelayedTask(
+          FROM_HERE, {content::BrowserThread::UI},
+          base::BindOnce(&BraveProfileSyncServiceImpl::OnResolvedPreferencesWrp,
```

Comment:
This is for the case when we want to reset the sync but internet connection is offline.
When the internet is offline, `FetchDevices` will get empty result, and we will not be able to call `ResetSyncInternal()`

Reply:
Actually we can simplify it by making sure DELETE device record is sent by `OnRecordsSent` and then we can call `ResetSyncInternal()` without involving `OnResolvedPreferences`

Reply:
This delayed task doesn't really solve this situation
```
IV. Check how works reset with Leave Sync Chain button when the device is offline

1. Create sync chain of 2 devices, deviceA and deviceB
2. Turn off the internet connection on deviceA
3. Press Leave Sync Chain button on deviceA
4. Ensure chain have been reset on deviceA, but there deviceB still have deviceA in list
5. Close browser deviceA; turn on the internet conenction, launch browser of deviceA - ensure there are no sync chain on deviceA
```
device B still doesn't get delete record from A

Here are what we can do
1. If we want to make sure device B gets DELETE A record, we need to block A from Reset until device A sends the record (which means device A get online)
2. Or we can leverage `OnRecordsSent` as I mentioned in https://github.com/brave/brave-core/pull/3988/files#r348770951
It will get called no matter success or failure, then we call `ResetSyncInternal()` (failure will have empty records)

Reply:
For the case when internet connection is off there could not be a perfect solution, because we cannot send device DELETE record.

The option with delayed task gives two good points:
1) If internet connection is on, then the device is able to send DELETE record, so it informs all other. No disadvantages here. 
2) If internet connection is off, then the device is able to make a proper self-reset. The only disadvantage is - other devices would not be informed and could see `zobmie`.

Any block of ourselves until get device DELETE record confirmation, makes sync to be under `pending reset` for an unknown time.

As for now, when the internet connection is off, sending DEVICE delete record will never reach to AWS, because we do resend for bookmarks only.

We can leverage `OnRecordsSent` and do there a cancellation of delayed task if we see device DELETE with our device_id. @darkdh what do you think?

Reply:
We don’t need delayed task if we use OnRecordsSent because when offline or any failed to send, it will get called with EMPTY records to indicate failure.

Reply:
As per Slack discussion, the case of reset when the device is offline will be done as separate issue (https://github.com/brave/brave-browser/issues/3628) + separete PR after this PR will be merged.